File: cancer_claw/resources/knowledge/skill_packs/bioSkills-main/workflows/timecourse-pipeline/examples/timecourse_pipeline.py
'''Time-course analysis pipeline: temporal DE, clustering, optional periodicity, GAM fitting, enrichment.'''
# Reference: gseapy 1.1+, numpy 1.26+, pandas 2.2+, pygam 0.9+, scikit-learn 1.4+, scipy 1.12+, statsmodels 0.14+, tslearn 0.6+ | Verify API if version differs

import pandas as pd
import numpy as np
from scipy import stats
from statsmodels.stats.multitest import multipletests
from patsy import dmatrix
from sklearn.metrics import silhouette_score
from tslearn.clustering import TimeSeriesKMeans
from pygam import LinearGAM, s
import gseapy as gp

# --- Configuration ---
COUNTS_FILE = 'counts_normalized.csv'
METADATA_FILE = 'metadata.csv'
OUTPUT_PREFIX = 'timecourse_results'
# FDR <0.05: standard threshold; relax to 0.1 for exploratory clustering
FDR_THRESHOLD = 0.05
# 4-20 clusters typical; start with 8, evaluate with silhouette
N_CLUSTERS = 8
# softdtw gamma: lower = stricter alignment; 0.01 works for most time courses
DTW_GAMMA = 0.01
# GAM n_splines: 5 default; reduce to 3 for <6 time points
N_SPLINES = 5
# Set True if experiment covers 24h+ cycles with 2-4h resolution
CIRCADIAN_DESIGN = False

# --- Step 1: Load data ---
expr = pd.read_csv(COUNTS_FILE, index_col=0)
meta = pd.read_csv(METADATA_FILE)
print(f'Loaded: {expr.shape[0]} genes x {expr.shape[1]} samples, {meta["time"].nunique()} time points')

# --- Step 2: Temporal DE (F-test on spline basis) ---
# df=3: cubic spline; increase to 4-5 for >10 time points
spline_basis = dmatrix('bs(time, df=3)', data=meta, return_type='dataframe')
design_full = np.column_stack([np.ones(len(meta)), spline_basis.values])
design_reduced = np.ones((len(meta), 1))
df_diff = design_full.shape[1] - design_reduced.shape[1]
df_resid = len(meta) - design_full.shape[1]

pvals = []
for gene in expr.index:
    y = expr.loc[gene].values
    beta_full = np.linalg.lstsq(design_full, y, rcond=None)[0]
    beta_red = np.linalg.lstsq(design_reduced, y, rcond=None)[0]
    ss_full = np.sum((y - design_full @ beta_full) ** 2)
    ss_red = np.sum((y - design_reduced @ beta_red) ** 2)
    f_stat = ((ss_red - ss_full) / df_diff) / (ss_full / df_resid)
    pvals.append(1 - stats.f.cdf(f_stat, df_diff, df_resid))

_, fdr, _, _ = multipletests(pvals, method='fdr_bh')
temporal_mask = fdr < FDR_THRESHOLD
temporal_genes = expr.index[temporal_mask].tolist()
print(f'Significant temporal genes (FDR <{FDR_THRESHOLD}): {len(temporal_genes)}')

# QC gate: sufficient temporal genes
if len(temporal_genes) < 100:
    print('WARNING: Few temporal genes detected. Check replicates or relax FDR.')

expr_sig = expr.loc[temporal_genes]

# --- Step 3: Time-series clustering (tslearn soft-DTW) ---
# Standardize per-gene (row-wise): z-score each gene's profile across timepoints
expr_scaled = (expr_sig.values - expr_sig.values.mean(axis=1, keepdims=True)) / expr_sig.values.std(axis=1, keepdims=True)

# Reshape for tslearn: (n_genes, n_timepoints, 1)
X = expr_scaled.reshape(expr_scaled.shape[0], expr_scaled.shape[1], 1)

model = TimeSeriesKMeans(n_clusters=N_CLUSTERS, metric='softdtw',
                         metric_params={'gamma': DTW_GAMMA},
                         max_iter=50, random_state=42)
labels = model.fit_predict(X)

cluster_sizes = pd.Series(labels).value_counts().sort_index()
print('Cluster sizes:')
print(cluster_sizes.to_string())

# QC gate: no empty clusters
if (cluster_sizes == 0).any():
    print('WARNING: Empty clusters found. Reduce N_CLUSTERS.')

# Silhouette score for cluster quality validation
sil = silhouette_score(expr_scaled, labels, metric='euclidean')
print(f'Mean silhouette score: {sil:.3f}')

cluster_df = pd.DataFrame({'gene': temporal_genes, 'cluster': labels})
cluster_df.to_csv(f'{OUTPUT_PREFIX}_clusters.csv', index=False)

# --- Step 4a: Optional periodicity detection (CosinorPy) ---
if CIRCADIAN_DESIGN:
    from cosinorpy import cosinor

    # fit_group expects long-format DataFrame with columns 'x' (time), 'y' (expression), 'test' (gene name)
    # Map sample names to time values from metadata before reshaping
    # period=24: standard circadian; adjust for ultradian (4-12h) or infradian (>28h)
    records = []
    for gene in expr_sig.index:
        for sample_idx, sample in enumerate(expr_sig.columns):
            records.append({'x': meta['time'].iloc[sample_idx], 'y': expr_sig.loc[gene, sample], 'test': gene})
    expr_long = pd.DataFrame(records)
    cosinor_results = cosinor.fit_group(expr_long, period=24, n_components=1)
    rhythmic = cosinor_results[cosinor_results['p'] < 0.05]
    print(f'Rhythmic genes (p <0.05): {len(rhythmic)}')
    rhythmic.to_csv(f'{OUTPUT_PREFIX}_rhythmic.csv', index=False)

# --- Step 4b: GAM trajectory fitting ---
time_vals = meta['time'].values.reshape(-1, 1)

for cl_id in range(N_CLUSTERS):
    cl_mask = labels == cl_id
    if cl_mask.sum() == 0:
        continue

    mean_profile = expr_scaled[cl_mask].mean(axis=0)
    gam = LinearGAM(s(0, n_splines=N_SPLINES)).fit(time_vals, mean_profile)
    print(f'Cluster {cl_id}: GAM GCV = {gam.statistics_["GCV"]:.4f}')

# --- Step 5: Per-cluster pathway enrichment (gseapy/Enrichr) ---
clusters_with_terms = 0

for cl_id in range(N_CLUSTERS):
    cl_genes = cluster_df[cluster_df['cluster'] == cl_id]['gene'].tolist()
    if len(cl_genes) < 5:
        print(f'Cluster {cl_id}: too few genes ({len(cl_genes)}), skipping enrichment')
        continue

    enr = gp.enrichr(gene_list=cl_genes, gene_sets='GO_Biological_Process_2023',
                     organism='human', outdir=f'{OUTPUT_PREFIX}_enrichr_cluster_{cl_id}')
    sig_terms = enr.results[enr.results['Adjusted P-value'] < 0.05]
    n_terms = len(sig_terms)
    if n_terms > 0:
        clusters_with_terms += 1
    print(f'Cluster {cl_id}: {n_terms} significant GO terms')

# QC gate: enrichment coverage
print(f'Clusters with significant GO terms: {clusters_with_terms} / {N_CLUSTERS}')
if clusters_with_terms < 3:
    print('WARNING: Few clusters enriched. Check gene naming convention or relax thresholds.')

print(f'\nPipeline complete: {len(temporal_genes)} temporal genes, {N_CLUSTERS} clusters, {clusters_with_terms} enriched')

File: cancer_claw/resources/knowledge/skill_packs/bioSkills-main/workflows/timecourse-pipeline/examples/timecourse_pipeline.R
# Reference: DESeq2 1.42+, clusterProfiler 4.10+, limma 3.58+, numpy 1.26+, pandas 2.2+, scanpy 1.10+, scipy 1.12+, statsmodels 0.14+ | Verify API if version differs
## Time-course analysis pipeline: temporal DE, Mfuzz clustering, optional MetaCycle,
## GAM trajectory fitting, and per-cluster pathway enrichment.

library(limma)
library(splines)
library(Mfuzz)
library(mgcv)
library(clusterProfiler)
library(org.Hs.eg.db)
library(cluster)

# --- Configuration ---
COUNTS_FILE <- 'counts_normalized.csv'
METADATA_FILE <- 'metadata.csv'
OUTPUT_PREFIX <- 'timecourse_results'
# FDR <0.05: standard temporal DE threshold; relax to 0.1 for exploratory analysis
FDR_THRESHOLD <- 0.05
# 4-20 clusters typical; start with 8, refine via gap statistic or silhouette
N_CLUSTERS <- 8
# Membership >0.5: core gene assignment; lower to 0.3 for exploratory
MIN_MEMBERSHIP <- 0.5
# GAM basis dimension; 5 default, reduce to 3 for <6 time points
GAM_K <- 5
# Set TRUE if experiment covers 24h+ cycles with 2-4h resolution
CIRCADIAN_DESIGN <- FALSE

# --- Step 1: Load data ---
expr <- as.matrix(read.csv(COUNTS_FILE, row.names = 1))
meta <- read.csv(METADATA_FILE)
message(sprintf('Loaded: %d genes x %d samples, %d time points',
                nrow(expr), ncol(expr), length(unique(meta$time))))

# --- Step 2: Temporal DE (limma splines) ---
# df=3: cubic spline sufficient for most time courses; increase to 4-5 for >10 time points
design <- model.matrix(~ ns(meta$time, df = 3))
fit <- lmFit(expr, design)
fit <- eBayes(fit)
temporal_results <- topTable(fit, coef = 2:ncol(design), number = Inf, sort.by = 'F')
# topTable already returns adj.P.Val (BH-corrected); use it directly
sig_genes <- rownames(temporal_results[temporal_results$adj.P.Val < FDR_THRESHOLD, ])
message(sprintf('Significant temporal genes (FDR <%s): %d', FDR_THRESHOLD, length(sig_genes)))

# QC gate: sufficient temporal genes
if (length(sig_genes) < 100) {
    message('WARNING: Few temporal genes detected. Check replicate count or relax FDR.')
}

expr_sig <- expr[sig_genes, ]

# --- Step 3: Mfuzz soft clustering ---
eset <- ExpressionSet(assayData = as.matrix(expr_sig))
eset <- standardise(eset)

# mestimate(): optimal fuzzifier from data geometry; typical range 1.5-2.5
m <- mestimate(eset)
message(sprintf('Estimated fuzzifier m = %.2f', m))

cl <- mfuzz(eset, c = N_CLUSTERS, m = m)

# QC gate: no empty clusters
cluster_sizes <- table(cl$cluster)
message('Cluster sizes:')
print(cluster_sizes)
if (any(cluster_sizes == 0)) {
    message('WARNING: Empty clusters found. Reduce N_CLUSTERS.')
}

core_genes <- acore(eset, cl, min.acore = MIN_MEMBERSHIP)
for (i in seq_along(core_genes)) {
    message(sprintf('Cluster %d: %d core genes (membership >%.1f)', i, nrow(core_genes[[i]]), MIN_MEMBERSHIP))
}

# Silhouette validation
sil <- silhouette(cl$cluster, dist(exprs(eset)))
message(sprintf('Mean silhouette score: %.3f', mean(sil[, 3])))

# Save cluster assignments
cluster_df <- data.frame(gene = names(cl$cluster), cluster = cl$cluster)
write.csv(cluster_df, paste0(OUTPUT_PREFIX, '_clusters.csv'), row.names = FALSE)

# --- Step 4a: Optional rhythm detection (MetaCycle) ---
if (CIRCADIAN_DESIGN) {
    library(MetaCycle)
    expr_for_meta <- expr_sig
    colnames(expr_for_meta) <- meta$time_hours
    write.csv(expr_for_meta, paste0(OUTPUT_PREFIX, '_for_metacycle.csv'))

    # Period range 20-28h: standard circadian window
    # Adjust to 4-12h for ultradian or >28h for infradian rhythms
    meta2d(paste0(OUTPUT_PREFIX, '_for_metacycle.csv'), filestyle = 'csv',
           minper = 20, maxper = 28,
           timepoints = sort(unique(meta$time_hours)),
           outdir = paste0(OUTPUT_PREFIX, '_metacycle'))
    message('MetaCycle rhythm detection complete.')
}

# --- Step 4b: GAM trajectory fitting ---
cluster_trajectories <- list()
for (cl_id in 1:N_CLUSTERS) {
    cl_gene_names <- names(cl$cluster[cl$cluster == cl_id])
    if (length(cl_gene_names) == 0) next

    mean_profile <- colMeans(expr_sig[cl_gene_names, , drop = FALSE])
    df_gam <- data.frame(time = meta$time, expr = mean_profile)

    gam_fit <- gam(expr ~ s(time, k = GAM_K), data = df_gam)
    r_sq <- summary(gam_fit)$r.sq
    edf <- summary(gam_fit)$edf

    cluster_trajectories[[cl_id]] <- list(fit = gam_fit, r_squared = r_sq, edf = edf)
    message(sprintf('Cluster %d: GAM R^2 = %.3f, EDF = %.2f', cl_id, r_sq, edf))
}

# --- Step 5: Per-cluster pathway enrichment ---
enrichment_results <- list()
clusters_with_terms <- 0

for (i in seq_along(core_genes)) {
    genes <- core_genes[[i]]$NAME
    if (length(genes) < 5) {
        message(sprintf('Cluster %d: too few genes (%d), skipping enrichment', i, length(genes)))
        next
    }

    entrez <- bitr(genes, fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = org.Hs.eg.db)

    # pvalueCutoff 0.05, qvalueCutoff 0.05: standard enrichment thresholds
    ego <- enrichGO(gene = entrez$ENTREZID, OrgDb = org.Hs.eg.db,
                    ont = 'BP', pAdjustMethod = 'BH',
                    pvalueCutoff = 0.05, qvalueCutoff = 0.05,
                    readable = TRUE)

    enrichment_results[[i]] <- ego
    n_terms <- nrow(as.data.frame(ego))
    if (n_terms > 0) clusters_with_terms <- clusters_with_terms + 1
    message(sprintf('Cluster %d: %d significant GO BP terms', i, n_terms))
}

# QC gate: enrichment coverage
message(sprintf('Clusters with significant GO terms: %d / %d', clusters_with_terms, N_CLUSTERS))
if (clusters_with_terms < 3) {
    message('WARNING: Few clusters enriched. Check gene ID mapping or relax thresholds.')
}

# --- Save enrichment results ---
for (i in seq_along(enrichment_results)) {
    if (!is.null(enrichment_results[[i]])) {
        df_enr <- as.data.frame(enrichment_results[[i]])
        if (nrow(df_enr) > 0) {
            write.csv(df_enr, sprintf('%s_cluster%d_GO.csv', OUTPUT_PREFIX, i), row.names = FALSE)
        }
    }
}

message(sprintf('\nPipeline complete: %d temporal genes, %d clusters, %d enriched',
                length(sig_genes), N_CLUSTERS, clusters_with_terms))
